botocore/signers.py

```diff
@@ -130,3 +134,275 @@ def get_auth(self, signing_name, region_name, signature_version=None):
+            of.
+
+        :type conditions: list
+        :param conditions: A list of conditions to include in the policy. Each
+            element can be either a list or a structure. For example:
+            [
+             {"acl": "public-read"},
+             {"bucket": "mybucket"},
+             ["starts-with", "$key", "mykey"]
+            ]
+
+        :type expires_in: int
+        :param expires_in: The number of seconds the presigned post is valid
+            for.
+
+        :type region_name: string
+        :param region_name: The region name to sign the presigned post to.
+
+        :rtype: dict
+        :returns: A dictionary with two elements: ``url`` and ``fields``.
+            Url is the url to post to. Fields is a dictionary filled with
+            the form fields and respective values to use when submitting the
+            post. For example:
+
+            {'url': 'https://mybucket.s3.amazonaws.com
+             'fields': {'acl': 'public-read',
+                        'key': 'mykey',
+                        'signature': 'mysignature',
+                        'policy': 'mybase64 encoded policy'}
+            }
+        """
+        if fields is None:
+            fields = {}
+
+        if conditions is None:
+            conditions = []
+
+        if region_name is None:
+            region_name = self._region_name
+
+        # Create the policy for the post.
+        policy = {}
+
+        # Create an expiration date for the policy
+        datetime_now = datetime.datetime.utcnow()
+        expire_date = datetime_now + datetime.timedelta(seconds=expires_in)
+        policy['expiration'] = expire_date.strftime(botocore.auth.ISO8601)
+
+        # Append all of the conditions that the user supplied.
+        policy['conditions'] = []
+        for condition in conditions:
+            policy['conditions'].append(condition)
+
+        # Obtain the appropriate signer.
+        query_prefix = '-presign-post'
+        signature_version = self._signature_version
+        if not signature_version.endswith(query_prefix):
+            signature_version += query_prefix
+
+        kwargs = {'signing_name': self._signing_name,
+                  'region_name': region_name,
+                  'signature_version': signature_version}
+        auth = self.get_auth(**kwargs)
+
+        # Store the policy and the fields in the request for signing
+        request = create_request_object(request_dict)
+        request.context['s3-presign-post-fields'] = fields
+        request.context['s3-presign-post-policy'] = policy
+
+        auth.add_auth(request)
+
+        # Fix s3 host for s3 sigv2 bucket names
+        signature_type = signature_version.split('-', 1)[0]
+        fix_s3_host(request, signature_type, region_name)
+        # Return the url and the fields for th form to post.
+        return {'url': request.url, 'fields': fields}
+
+
+def generate_url(client, client_method, client_kwargs=None, expires_in=3600):
```

Comment:
Would would you think about allowing the user to provide the HTTP method?

I could someone generating a presigned url for an operation that's normally a POST, but wanting to just curl the url with a GET.